Zend/zend_hash.c

```diff
@@ -480,6 +480,31 @@ ZEND_API void ZEND_FASTCALL _zend_hash_iterators_update(HashTable *ht, HashPosit
 	}
 }
 
+/* The macro casts strings to be compared to zend_long arrays truncating the length to multiple of
+ * zend_long size. This is acceptable for hash string comparisons as the ignored tailing bytes are already
+ * implicitly checked by hash value matching.
+ * Hash value function ensures (see zend_inline_hash_func() in zend_string.h):
+ * - strings with lengths < SIZEOF_ZEND_LONG will generate unique hash values
+ * - strings with identical length differing only in last (SIZEOF_ZEND_LONG-1) bytes will always generate
+ *   different hash values.
+ */
+#define ZEND_STR2LONG_MEMEQ(s1, s2 ,len ) \
+		zend_ulong_memeq((const zend_ulong *)(s1), (const zend_ulong *)(s2), (len) >> (SIZEOF_ZEND_LONG_LOG2))
+
+/* returns 1 if the zend_log arrays are equal, 0 otherwise; empty arrays are considered equal. */
+static zend_always_inline zend_bool zend_ulong_memeq(const zend_ulong *arr1, const zend_ulong *arr2, size_t len)
+{
+	while (len != 0) {
+		if ( *arr1 != *arr2 ) {
+			return 0;
+		}
+		arr1++;
+		arr2++;
+		len --;
+	}
+	return 1; /* equal arrays */
+}
```

Comment:
I think such optimization already be done by compiler(some even with SIMD), did you see any difference via perf tool?